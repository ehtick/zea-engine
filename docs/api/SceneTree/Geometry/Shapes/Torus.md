<a name="Torus"></a>

### Torus 
A class for generating a torus geometry.

```
const torus = new Torus(0.4, 1.3)
```


**Extends**: <code>[ProceduralMesh](api/SceneTree/Geometry/Shapes/ProceduralMesh.md)</code>  
<a name="new_Torus_new"></a>

### new Torus
Creates an instance of Torus.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| [innerRadius] | <code>number</code> | <code>0.5</code> | The inner radius of the torus. |
| [outerRadius] | <code>number</code> | <code>3</code> | The outer radius of the torus. |
| [detail] | <code>number</code> | <code>32</code> | The detail of the cone. |
| [loops] | <code>number</code> | <code>6</code> | The number of loops. |



### [Class Tests](api/SceneTree/Geometry/Shapes/Torus.test)