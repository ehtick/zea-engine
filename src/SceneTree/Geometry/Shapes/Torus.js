import { Vec2 } from '../../../Math/Vec2'
import { Vec3 } from '../../../Math/Vec3'
import { Mesh } from '../Mesh.js'

/**
 * A class for generating a torus geometry.
 *
 * ```
 * const torus = new Torus(0.4, 1.3)
 * ```
 *
 * @extends Mesh
 */
class Torus extends Mesh {
  /**
   * Creates an instance of Torus.
   *
   * @param {number} [innerRadius=0.5] - The inner radius of the torus.
   * @param {number} [outerRadius=1.0] - The outer radius of the torus.
   * @param {number} [detail=32] - The detail of the cone.
   */
  constructor(innerRadius = 0.5, outerRadius = 1.0, detail = 32) {
    super()

    if (isNaN(innerRadius) || isNaN(outerRadius) || isNaN(detail)) throw new Error('Invalid geom args')

    this.__innerRadius = innerRadius
    this.__outerRadius = outerRadius
    this.__detail = detail >= 3 ? detail : 3

    this.addVertexAttribute('texCoords', Vec2)
    this.addVertexAttribute('normals', Vec3)
    this.__rebuild()
  }

  /**
   * Getter for the inner radius.
   *
   * @return {number} - Returns the radius.
   */
  get innerRadius() {
    return this.__innerRadius
  }

  /**
   * Setter for the inner radius.
   *
   * @param {number} val - The radius value.
   */
  set innerRadius(val) {
    this.__innerRadius = val
    this.__resize()
  }

  /**
   * Getter for the outer radius.
   *
   * @return {number} - Returns the radius.
   */
  get outerRadius() {
    return this.__outerRadius
  }

  /**
   * Setter for the outer radius.
   *
   * @param {number} val - The radius value.
   */
  set outerRadius(val) {
    this.__outerRadius = val
    this.__resize()
  }

  /**
   * Getter for the torus detail.
   *
   * @return {number} - Returns the detail.
   */
  get detail() {
    return this.__detail
  }

  /**
   * Setter for the torus detail.
   *
   * @param {number} val - The detail value.
   */
  set detail(val) {
    this.__detail = val >= 3 ? val : 3
    this.__rebuild()
  }

  /**
   * The __rebuild method.
   * @private
   */
  __rebuild() {
    const nbSlices = this.__detail
    const nbLoops = this.__detail * 2
    const numVertices = nbSlices * nbLoops

    this.setNumVertices(numVertices)
    this.setFaceCounts([0, nbSlices * nbLoops])

    // ////////////////////////////
    // Set Vertex Positions

    const normals = this.getVertexAttribute('normals')
    let vertex = 0
    for (let i = 0; i < nbLoops; i++) {
      const theta = (i / nbLoops) * 2.0 * Math.PI
      const ctheta = Math.cos(theta)
      const stheta = Math.sin(theta)

      for (let j = 0; j < nbSlices; j++) {
        const phi = (j / nbSlices) * 2.0 * Math.PI

        const sphi = Math.sin(phi)
        const cphi = Math.cos(phi)
        const d = this.__outerRadius + cphi * this.__innerRadius

        // Set positions and normals at the same time.
        this.getVertex(vertex).set(ctheta * d, stheta * d, this.__innerRadius * sphi)
        normals.getValueRef(vertex).set(ctheta * cphi, stheta * cphi, sphi)
        vertex++
      }
    }

    // ////////////////////////////
    // Build the topology and texCoords
    const texCoords = this.getVertexAttribute('texCoords')
    let faceIndex = 0
    for (let i = 0; i < nbLoops; i++) {
      for (let j = 0; j < nbSlices; j++) {
        const ip = (i + 1) % nbLoops
        const jp = (j + 1) % nbSlices
        const v0 = nbSlices * i + j
        const v1 = nbSlices * i + jp
        const v2 = nbSlices * ip + jp
        const v3 = nbSlices * ip + j
        this.setFaceVertexIndices(faceIndex, v0, v1, v2, v3)

        texCoords.setFaceVertexValue(faceIndex, 0, new Vec2(i / nbLoops, j / nbLoops))
        texCoords.setFaceVertexValue(faceIndex, 1, new Vec2(i / nbLoops, (j + 1) / nbLoops))
        texCoords.setFaceVertexValue(faceIndex, 2, new Vec2((i + 1) / nbLoops, (j + 1) / nbLoops))
        texCoords.setFaceVertexValue(faceIndex, 3, new Vec2((i + 1) / nbLoops, j / nbLoops))
        faceIndex++
      }
    }

    this.setBoundingBoxDirty()
    this.emit('geomDataTopologyChanged', {})
  }

  /**
   * The __resize method.
   * @private
   */
  __resize() {
    const nbSlices = this.__detail
    const nbLoops = this.__detail * 2

    const vertex = 0
    for (let i = 0; i < nbLoops; i++) {
      const theta = (i / nbLoops) * 2.0 * Math.PI
      const ctheta = Math.cos(theta)
      const stheta = Math.sin(theta)

      for (let j = 0; j < nbSlices; j++) {
        const phi = (j / nbSlices) * 2.0 * Math.PI

        const sphi = Math.sin(phi)
        const cphi = Math.cos(phi)
        const d = this.__outerRadius + cphi * this.__innerRadius

        // Set positions and normals at the same time.
        this.getVertex(vertex).set(ctheta * d, stheta * d, this.__innerRadius * sphi)
        index++
      }
    }

    this.setBoundingBoxDirty()
    this.emit('geomDataChanged', {})
  }

  /**
   * The toJSON method encodes this type as a json object for persistence.
   * @return {object} - Returns the json object.
   */
  toJSON() {
    const json = super.toJSON()
    json['x'] = this.__x
    json['y'] = this.__y
    json['z'] = this.__z
    return json
  }
}

export { Torus }
