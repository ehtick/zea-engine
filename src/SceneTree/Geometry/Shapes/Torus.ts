import { Vec2 } from '../../../Math/Vec2'
import { ProceduralMesh } from './ProceduralMesh'
import { Registry } from '../../../Registry'
import { NumberParameter } from '../../Parameters/NumberParameter'
import { Vec3Attribute } from '../Vec3Attribute'
import { Vec2Attribute } from '../Vec2Attribute'

/**
 * A class for generating a torus geometry.
 *
 * ```
 * const torus = new Torus(0.4, 1.3)
 * ```
 *
 * @extends ProceduralMesh
 */
class Torus extends ProceduralMesh {
  /**
   * @member {NumberParameter} arcAngleParam - TODO
   */
  arcAngleParam: NumberParameter

  /**
   * @member {NumberParameter} detailParam - TODO
   */
  detailParam: NumberParameter

  /**
   * @member {NumberParameter} innerRadiusParam - TODO
   */
  innerRadiusParam: NumberParameter

  /**
   * @member {NumberParameter} outerRadiusParam - TODO
   */
  outerRadiusParam: NumberParameter

  /**
   * Creates an instance of Torus.
   *
   * @param {number} [innerRadius=0.5] - The inner radius of the torus.
   * @param {number} [outerRadius=3] - The outer radius of the torus.
   * @param {number} [detail=32] - The detail of the cone.
   * @param {number} [arcAngle=Math.PI * 2.0] - The angle of the arc.
   */
  constructor(innerRadius = 0.5, outerRadius = 3, detail = 32, arcAngle = Math.PI * 2.0) {
    super()

    if (isNaN(innerRadius) || isNaN(outerRadius) || isNaN(detail)) throw new Error('Invalid geom args')

    this.innerRadiusParam = this.addParameter(new NumberParameter('InnerRadius', innerRadius)) as NumberParameter
    this.outerRadiusParam = this.addParameter(new NumberParameter('OuterRadius', outerRadius)) as NumberParameter
    this.detailParam = this.addParameter(
      new NumberParameter('Detail', detail >= 3 ? detail : 3, [3, 200], 1)
    ) as NumberParameter
    this.arcAngleParam = this.addParameter(new NumberParameter('ArcAngle', arcAngle)) as NumberParameter

    this.addVertexAttribute('texCoords', new Vec2Attribute())
    this.addVertexAttribute('normals', new Vec3Attribute())

    this.topologyParams.push('Detail')
    this.topologyParams.push('ArcAngle')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const arcAngle = this.arcAngleParam.value || Math.PI * 2.0
    const open = arcAngle < 2.0 * Math.PI
    const detail = this.detailParam.value || 32
    const nbSlices = detail
    const nbLoops = detail * 2 + (open ? 1 : 0)
    const numVertices = nbSlices * nbLoops

    this.setNumVertices(numVertices)
    this.setFaceCounts([0, nbSlices * nbLoops])

    // ////////////////////////////
    // Build the topology and texCoords
    const texCoords = <Vec2Attribute>this.getVertexAttribute('texCoords')
    if (texCoords) {
      let faceIndex = 0
      for (let i = 0; i < (open ? nbLoops - 1 : nbLoops); i++) {
        for (let j = 0; j < nbSlices; j++) {
          const ip = (i + 1) % nbLoops
          const jp = (j + 1) % nbSlices
          const v0 = nbSlices * i + j
          const v1 = nbSlices * i + jp
          const v2 = nbSlices * ip + jp
          const v3 = nbSlices * ip + j
          this.setFaceVertexIndices(faceIndex, [v0, v1, v2, v3])

          texCoords.setFaceVertexValue(faceIndex, 0, new Vec2(i / nbLoops, j / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 1, new Vec2(i / nbLoops, (j + 1) / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 2, new Vec2((i + 1) / nbLoops, (j + 1) / nbLoops))
          texCoords.setFaceVertexValue(faceIndex, 3, new Vec2((i + 1) / nbLoops, j / nbLoops))
          faceIndex++
        }
      }
    }

    this.resize()
  }

  /**
   * The resize method.
   * @private
   */
  resize(): void {
    const innerRadius = this.innerRadiusParam.value || 0.5
    const outerRadius = this.outerRadiusParam.value || 3
    const arcAngle = this.arcAngleParam.value || Math.PI * 2
    const detail = this.detailParam.value || 32
    const open = arcAngle < 2.0 * Math.PI
    const nbSlices = detail
    const nbLoops = detail * 2 + (open ? 1 : 0)

    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    const normals = <Vec3Attribute>this.getVertexAttribute('normals')
    if (!positions || !normals) return

    let vertex = 0
    for (let i = 0; i < nbLoops; i++) {
      // const theta = (i / nbLoops) * arcAngle
      const theta = -((i / (open ? nbLoops - 1 : nbLoops)) * arcAngle)
      const ctheta = Math.cos(theta)
      const stheta = Math.sin(theta)

      for (let j = 0; j < nbSlices; j++) {
        const phi = (j / nbSlices) * 2.0 * Math.PI

        const sphi = Math.sin(phi)
        const cphi = Math.cos(phi)
        const d = outerRadius + cphi * innerRadius

        // Set positions and normals at the same time.
        positions.getValueRef(vertex).set(ctheta * d, stheta * d, innerRadius * sphi)
        normals.getValueRef(vertex).set(ctheta * cphi, stheta * cphi, sphi)
        vertex++
      }
    }
  }
}

Registry.register('Torus', Torus)

export { Torus }
