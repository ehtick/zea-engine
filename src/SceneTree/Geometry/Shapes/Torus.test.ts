import { Torus } from './Torus'

describe('Torus', () => {
  test('test default parameters.', () => {
    const torus = new Torus()

    expect(torus.getParameter('InnerRadius').getValue()).toBe(0.5)

    expect(torus.getParameter('OuterRadius').getValue()).toBe(3)

    expect(torus.getParameter('Detail').getValue()).toBe(32)
  })

  test('test update parameters.', () => {
    const torus = new Torus()
    torus.getParameter('InnerRadius').setValue(3)
    torus.getParameter('OuterRadius').setValue(5)
    torus.getParameter('Detail').setValue(16)

    expect(torus.getParameter('InnerRadius').getValue()).toBe(3)

    expect(torus.getParameter('OuterRadius').getValue()).toBe(5)

    expect(torus.getParameter('Detail').getValue()).toBe(16)
  })

  it('saves to JSON (serialization).', () => {
    const torus = new Torus(3, 5, 8)

    const outputJSON = torus.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const torus = new Torus()
    const inputJSON = {
      params: {
        Detail: {
          range: [3, 200],
          step: 1,
          value: 8,
        },
        InnerRadius: {
          value: 3,
        },
        OuterRadius: {
          value: 5,
        },
      },
      type: 'Torus',
      vertexAttributes: {},
    }

    torus.fromJSON(inputJSON)

    expect(torus.toJSON()).toEqual(new Torus(3, 5, 8).toJSON())
  })
})
