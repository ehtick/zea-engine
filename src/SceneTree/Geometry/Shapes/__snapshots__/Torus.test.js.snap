// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Torus saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "Detail": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 8,
    },
    "InnerRadius": Object {
      "value": 3,
    },
    "Loops": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 6,
    },
    "OuterRadius": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 5,
    },
  },
  "type": "Torus",
  "vertexAttributes": Object {},
}
`;
